// frontend/src/context/InventoryContext.jsx
import React, { createContext, useState, useEffect, useContext } from 'react';
import { io } from 'socket.io-client';
import { AuthContext } from './AuthContext';

export const InventoryContext = createContext();

const API_URL = import.meta.env.VITE_API_URL || '/api';
const SOCKET_URL = import.meta.env.VITE_SOCKET_URL || window.location.origin;

export const InventoryProvider = ({ children }) => {
  const [products, setProducts] = useState([]);
  const [orders, setOrders] = useState([]);
  const [warehouses, setWarehouses] = useState([]);
  const [socketConnected, setSocketConnected] = useState(false);
  const { user } = useContext(AuthContext);

  const fetchProducts = async () => {
    if (!user) return;
    try {
      const res = await fetch(`${API_URL}/products`, {
        headers: { Authorization: `Bearer ${user.token}` }
      });
      if (res.ok) setProducts(await res.json());
    } catch (error) {
      console.error('Error fetching products:', error);
    }
  };

  const fetchOrders = async () => {
    if (!user) return;
    try {
      const res = await fetch(`${API_URL}/orders`, {
        headers: { Authorization: `Bearer ${user.token}` }
      });
      if (res.ok) setOrders(await res.json());
    } catch (error) {
      console.error('Error fetching orders:', error);
    }
  };

  const fetchWarehouses = async () => {
    if (!user) return;
    try {
      const res = await fetch(`${API_URL}/warehouses`, {
        headers: { Authorization: `Bearer ${user.token}` }
      });
      if (res.ok) setWarehouses(await res.json());
    } catch (error) {
      console.error('Error fetching warehouses:', error);
    }
  };

  useEffect(() => {
    if (user) {
      fetchProducts();
      fetchOrders();
      fetchWarehouses();

      const socket = io(SOCKET_URL, {
        transports: ['websocket', 'polling'],
        reconnection: true,
        reconnectionDelay: 1000,
        reconnectionDelayMax: 5000,
        reconnectionAttempts: 5
      });
      
      socket.on('connect', () => {
        console.log('Socket connected');
        setSocketConnected(true);
      });
      
      socket.on('disconnect', () => {
        console.log('Socket disconnected');
        setSocketConnected(false);
      });
      
      socket.on('product-created', fetchProducts);
      socket.on('product-updated', fetchProducts);
      socket.on('order-created', fetchOrders);
      socket.on('stock-changed', () => { 
        fetchProducts(); 
        fetchOrders(); 
      });

      socket.on('error', (error) => {
        console.error('Socket error:', error);
      });

      return () => socket.disconnect();
    }
  }, [user]);

  return (
    <InventoryContext.Provider value={{
      products, orders, warehouses, fetchProducts, fetchOrders, fetchWarehouses, socketConnected
    }}>
      {children}
    </InventoryContext.Provider>
  );
};
